<launch>
  <!-- Basic parameters -->
  <param name="use_sim_time" type="bool" value="True"/>
  <!-- <arg name="rtabmap_args" default="  ddddelete_db_on_start"> -->

  <arg name="localization" default="true"/>
  <arg name="rtabmap_args" default="--database_path /home/fizzer/.ros/rtabmap.db"/>


  <!-- Static transform for base_footprint to RGB-D camera link (adjusted height) -->
  <node pkg="tf" type="static_transform_publisher" name="base_to_camera_tf" args="0 0 0 1.570796327 3.141592654 0.5207963268 base_footprint_custom camera_link 5" />
  <!-- <node pkg="tf" type="static_transform_publisher" name="imu_to_base_tf" args="0 0 0 0 0 0 imu_link base_link 100" /> -->
  <node pkg="tf" type="static_transform_publisher" name="reset_map_to_odom" args="0 0 0 0 0 0 map odom_custom 100" />

   <!-- Odometry Publisher Node -->
  <node name="odometry_publisher" pkg="robot_controller" type="odometry_publisher.py" output="screen" />

  
  <!-- RGB-D Mapping Configuration for RTAB-Map -->
  <node name="rtabmap" pkg="rtabmap_slam" type="rtabmap" output="screen" args="$(arg rtabmap_args)">
      <param name="frame_id" type="string" value="base_footprint_custom"/>
      <param name="odom_frame_id" type="string" value="odom_custom"/>
      <param name="wait_for_transform" type="bool" value="true"/>
      <param name="queue_size" type="int" value="100"/>
      <param name="approx_sync" type="bool" value="true"/>

      <!-- Makes sure the map isn't overriden -->
      <param name="Mem/IncrementalMemory" type="bool" value="false"/>

      <!-- VISUAL-BASED ENHANCEMENTS -->
      <!-- Parameters to improve reliance on visual data -->
      <param name="RGBD/LinearUpdate" type="double" value="0.1"/> <!-- More frequent updates -->
      <param name="RGBD/AngularUpdate" type="double" value="0.1"/> <!-- More frequent updates -->
      <param name="Mem/RehearsalSimilarity" type="double" value="0.1"/> <!-- Loosen similarity threshold -->
      <param name="RGBD/ProximityBySpace" type="bool" value="true"/> <!-- Enable proximity detection by appearance -->
      <param name="RGBD/OptimizeMaxError" type="double" value="10.0"/> <!-- Relax max error for optimization -->
      <param name="Kp/MaxFeatures" type="int" value="500"/> <!-- Increase feature count -->
      <param name="RGBD/OptimizeFromGraphEnd" type="bool" value="true"/> <!-- Optimize map correction -->
      <param name="RGBD/CreateOccupancyGrid" type="bool" value="true"/>
      <param name="RGBD/CloudMaxDepth" type="double" value="5.0"/>
      <!-- End of VISUAL-BASED ENHANCEMENTS -->

      <param name="RGBD/LoopClosureHypothesis" type="double" value="0.001"/> 
      <param name="Vis/MinInliers" type="int" value="6"/> <!-- Default is 15 -->
      <param name="RGBD/ProximityMaxGraphDepth" type="int" value="20"/>
      <param name="RGBD/ProximityPathMaxNeighbors" type="int" value="50"/>
      <param name="RGBD/OptimizeGravity" type="bool" value="false"/>

      <param name="publish_tf" type="bool" value="false"/>

      <param name="Mem/ResetCurrentPos" type="bool" value="true"/>
      <param name="Mem/InitWMWithAllNodes" type="bool" value="false"/>
      <param name="initial_pose" type="string" value="0.0 0.0 0.0 0.0 0.0 3.14159"/>


      <param name="Mem/LocalizationDataSaved" type="bool" value="false"/>
      <param name="RGBD/StartAtOrigin" type="bool" value="true"/>


      <!-- Use separate RGB and Depth topics (disable combined rgbd_image topic) -->
      <param name="subscribe_rgbd" type="bool" value="false"/>
      <param name="subscribe_depth" type="bool" value="true"/>
      <param name="subscribe_rgb" type="bool" value="true"/>
      <param name="RGBD/Enabled" type="bool" value="true"/>

      <!-- Remap RGB and Depth topics separately -->
      <remap from="rgb/image" to="/B1/rgbd_camera/rgb/image_raw"/>
      <remap from="depth/image" to="/B1/rgbd_camera/depth/image_raw"/>
      <remap from="rgb/camera_info" to="/B1/rgbd_camera/rgb/camera_info"/>
      <remap from="depth/camera_info" to="/B1/depth_camera/depth/camera_info"/>

  </node>

  <!-- Visualization Configuration for RTAB-Map -->
  <node pkg="rtabmap_viz" type="rtabmap_viz" name="rtabmap_viz" output="screen">
    <param name="subscribe_depth" type="bool" value="true"/>
    <param name="frame_id" type="string" value="base_footprint_custom"/>
    <param name="wait_for_transform" type="bool" value="true"/>
    <param name="queue_size" type="int" value="30"/>


    <!-- Remap RGB-D topics for visualization without combined rgbd_image topic -->
    <remap from="rgb/image" to="/B1/rgb_camera/rgb/image_raw"/>
    <remap from="depth/image" to="/B1/depth_camera/depth/image_raw"/>
    <remap from="rgb/camera_info" to="/B1/rgb_camera/rgb/camera_info"/>
    <remap from="depth/camera_info" to="/B1/depth_camera/depth/camera_info"/>
    <remap from="/odom" to="/odom_custom"/>
  </node>

  <!-- Optional: RVIZ for additional visualization (uncomment if needed) -->
  
  <!-- <node pkg="rviz" type="rviz" name="rviz" args="-d $(find rtabmap_demos)/launch/config/demo_robot_mapping.rviz" output="screen"/> -->
 
</launch>